proposed/middleware.md

```diff
@@ -0,0 +1,103 @@
+HTTP Middleware
+===============
+
+This document describes a common standard for HTTP middleware components using
+[PSR-7][psr7] compliant HTTP messages.
+
+HTTP middleware has been an important concept on other web development platforms
+for a good number of years, and since the inception of PSR-7 has been growing
+increasingly popular in the PHP world.
+
+An informal (de-facto) standard for the signature of middleware components is
+already widely in use, and this PSR serves to provide a formal definition in
+the form of a common standard interface for middleware components.
+
+The interface described in this document is an abstraction of an individual
+middleware component.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119][rfc2119].
+
+[psr7]: http://www.php-fig.org/psr/psr-7/
+[rfc2119]: http://tools.ietf.org/html/rfc2119
+
+## 1. Specification
+
+An HTTP middleware component is an individual component participating, together
+with other middleware components, in the processing of an incoming HTTP Request
+and the creation of a resulting HTTP Response, as defined by PSR-7.
+
+Middleware components MUST implement this interface.
+
+Middleware consumers (e.g. frameworks and middleware stacks) MUST type-hint any
+method accepting middleware components as arguments formally as `callable`, and
+informally as `Psr\Http\Middleware\MiddlewareInterface`, e.g. using php-doc tags:
+
+```php
+/**
+ * @param MiddlewareInterface $middleware
+ */
+public function push(callable $middleware)
+{
+    // ...
+}
+```
+
+## 2. Interfaces
+
+The following interface MUST be implemented by any compatible middleware component.
+
+
+```php
+namespace Psr\Http\Middleware;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+interface MiddlewareInterface
+{
+    /**
+     * Process a request and return a response.
+     *
+     * Takes the incoming request and optionally modifies it before delegating
+     * to the next handler to get a response. May modify the response before
+     * ultimately returning it.
+     *
+     * @param RequestInterface $request
+     * @param ResponseInterface $response
+     * @param callable $next delegate function that will dispatch the next middleware component:
+     *  function (RequestInterface $request, ResponseInterface $response): ResponseInterface
+     *
+     * @return ResponseInterface
+     */
+    public function __invoke(
+        RequestInterface $request,
+        ResponseInterface $response,
+        callable $next
+    );
+}
+```
+
+Note that a dispatcher MUST NOT pass an actual middleware component as the
+callable `$next` - this callable is not the actual middleware component, but
+a *delegate* function to dispatch the next middleware component, generated by
+the dispatcher. A simple implementation MAY be defined as:
```

Comment:
I would change this from `MAY be defined` to `could be defined` (lower case `could`) since this is merely a suggestion/example, and not a rule or recommendation pertaining to the RFC per se.